File: 2GII/PDOO/QytetetJava/src/modeloqytetet/Especulador.java
package modeloqytetet;

public class Especulador extends Jugador {
    private int fianza;
    
    // ----------------------- Constructor -------------------------------------   
    protected Especulador(Jugador jugador, int fianza) {
        super(jugador);
        this.fianza = fianza;
        factorEspeculador = 2;
    } 
    
    // ------------------------- Consultor -------------------------------------   
    
    // ------------------------- Métodos  --------------------------------------   
    @Override
    protected void pagarImpuestos(int cantidad) {
        modificarSaldo(-(cantidad/2)); 
    }
    
    private Boolean pagarFianza(int cantidad) {
        Boolean puedoPagar = false;
        
        if (super.tengoSaldo(cantidad)) {
            super.modificarSaldo(-cantidad);
            puedoPagar = true;
        }
        
        return puedoPagar;
    }
    
    @Override
    protected void irACarcel(Casilla casilla) {
        if (!pagarFianza(fianza))
            super.irACarcel(casilla);
    }
    
    @Override
    protected Especulador convertirme(int fianza) {
        return this;
    }
    
    @Override
    public String toString() {
        String result = super.toString();
        
        return result + "$$ Especulador $$";
    }
}

File: 2GII/PDOO/QytetetRuby/lib/especulador.rb
require_relative "./jugador"

module ModeloQytetet
  
  class Especulador < Jugador
    
    # -------------------------- Constructor -----------------------------------    
    def initialize (jugador, fianza)
      # Si por algún casual tambien tenías: super.nuevoJugador(...)
      # super llama siempre al método con el mismo nombre pero en la clase
      # padre. No es lo mismo que el puntero "this" de Java o C++
      nuevoJugador(jugador.clone)            
      
      @fianza = fianza
      @@factor_especulador = 2
    end
    
    # ---------------------------- Métodos -------------------------------------
    def pagar_impuestos(cantidad)
      super.modificar_saldo(-(cantidad/2))
    end
    
    def pagar_fianza(cantidad)
      puego_pagar = false
      
      if (super.tengo_saldo(cantidad))
        super.modificar_saldo(-cantidad)
        puedo_pagar = true
      end
      
      puedo_pagar
    end
    
    def ir_a_carcel(casilla)
      if(!pagar_fianza(fianza))
        super.ir_a_carcel(casilla)
      end
    end
    
    def convertirme(fianza)
      return self
    end
    
    def to_s
      super.to_s + "$$ Especulador $$"
    end
    
    protected :pagar_impuestos, :ir_a_carcel
    private :pagar_fianza
  end
  
end
